https://issues.redhat.com/browse/ACM-30219 Add ARO CAPZ doc (#8616)
open-cluster-management/rhacm-docs: clusters/cluster_api/create_capz_aro.adoc

:_mod-docs-content-type: PROCEDURE

[create-aro-hcp-capz]
= Creating Azure Red Hat OpenShift {hcp} clusters by using Cluster API Provider for Azure

[role="_abstract"]

> Create an Azure Red Hat OpenShift {hcp} managed cluster by using Cluster API Provider for Azure to reduce costs and setup time by swapping dedicated master VMs for a managed, containerized control plane. Use a GitOps workflow to provision and scale clusters by using the same declarative manifests that you use for your applications.

Reviewer: We do not typically abbreviate "virtual machines". Also is it possible to say "main virtual machines" instead of "master virtual machines". When you mention GitOps, are you referring to the framework or are we referring to Red Hat OpenShift GitOps?
```suggestion
Create an Azure Red Hat OpenShift {hcp} managed cluster by using Cluster API Provider for Azure to reduce costs and setup time by swapping dedicated main virtual machines for a managed, containerized control plane. Use a GitOps workflow to provision and scale clusters by using the same declarative manifests that you use for your applications.
```

Author: The framework in general, I'll also change it to main virtual machines

*Required access:* Cluster administrator

.Prerequisites

> - You must complete the preparation procedure. See xref:../cluster_api/prepare_capz_aro.adoc##prepare-create-aro-hcp-capz[Preparing to create Azure Red Hat OpenShift {hcp} clusters by using Cluster API Provider for Azure] to learn more.

Reviewer: ```suggestion
- You must complete the preparation procedure. See xref:../cluster_api/prepare_capz_aro.adoc#prepare-create-aro-hcp-capz[Preparing to create Azure Red Hat OpenShift {hcp} clusters by using Cluster API Provider for Azure] to learn more.
```

Author: Good catch!

> - You need the `clusterctl` tool. See link:https://docs.redhat.com/en/documentation/openshift_container_platform/4.21/html/cli_tools/openshift-cli-oc[OpenShift CLI (oc)] in the {ocp-short} documentation to learn more.

Reviewer 2: There is no info about the clusterctl tool on the link. The link is for `oc` cli tool.
Let's use https://cluster-api.sigs.k8s.io/clusterctl/overview instead.

Author: Updated

  location: eastus
  resourceGroupName: aro-hcp-cluster-rg
  subscriptionID: "<subscription_id>"
  resources:
---
apiVersion: controlplane.cluster.x-k8s.io/v1beta2
kind: AROControlPlane
metadata:
  name: aro-hcp-cluster
  namespace: aro-clusters
spec:
  location: eastus
  subscriptionID: "<subscription_id>"
  version: "4.21.0"
  resources:
---
apiVersion: cluster.x-k8s.io/v1beta2
kind: MachinePool
metadata:
  name: aro-hcp-cluster-workers
  namespace: aro-clusters
spec:
  clusterName: aro-hcp-cluster
  replicas: 3
  template:
    spec:
      bootstrap:
        dataSecretName: ""
      clusterName: aro-hcp-cluster
      infrastructureRef:
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
        kind: AROMachinePool
        name: aro-hcp-cluster-workers
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
kind: AROMachinePool
metadata:
  name: aro-hcp-cluster-workers
  namespace: aro-clusters
spec:
  resources:
----

+
- Replace `<subscription_id>` with your subscription ID.
- Define your infrastructure resources in `spec.resources` field of the `AROCluster` resource.
- Define your control plane resources in `spec.resources` field of the `AROControlPlane` resource.

> - Define your node pool resources in `spec.resources` field of the `AROMachinePool` resource.

Reviewer: This YAML is pretty long. Can it be split into smaller pieces? Are all of the contents in Replace `<subscription_id>` with your subscription **one** YAML file? I recommend that you create a step for each YAML file snippet. For example:

```suggestion
. Create a YAML file named `aro-hcp-cluster.yaml` to define your cluster resources. Replace `<subscription_id>` with your subscription ID. Your resource might resemble the following example:

+
[source,yaml]
----
apiVersion: cluster.x-k8s.io/v1beta2
kind: Cluster
metadata:
  name: aro-hcp-cluster
  namespace: aro-clusters
  labels:
    cluster.x-k8s.io/cluster-name: aro-hcp-cluster
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
        - 10.128.0.0/14
    services:
      cidrBlocks:
        - 172.30.0.0/16
  controlPlaneRef:
    apiVersion: controlplane.cluster.x-k8s.io/v1beta2
    kind: AROControlPlane
    name: aro-hcp-cluster
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
    kind: AROCluster
    name: aro-hcp-cluster
----

. Create an `AROCluster` resource and define your infrastructure resources in `spec.resources`. Replace `<subscription_id>` with your subscription ID. Your resource might resemble the following example:

+
[source,bash]
----
apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
kind: AROCluster
metadata:
  name: aro-hcp-cluster
  namespace: aro-clusters
spec:
  location: eastus
  resourceGroupName: aro-hcp-cluster-rg
  subscriptionID: "<subscription_id>"
  resources:
----

. Create an `AROControlPlane` resource and define your control plane resources in `spec.resources` field. Your resource might resemble the following example:

+
[source,bash]
----
apiVersion: controlplane.cluster.x-k8s.io/v1beta2
kind: AROControlPlane
metadata:
  name: aro-hcp-cluster
  namespace: aro-clusters
spec:
  location: eastus
  subscriptionID: "<subscription_id>"
  version: "4.21.0"
  resources:
----

. Create a `MachinePool` resource and define your node pool resources in `spec.resources` field. Your resource might resemble the following example:

+
[source,bash]
----
apiVersion: cluster.x-k8s.io/v1beta2
kind: MachinePool
metadata:
  name: aro-hcp-cluster-workers
  namespace: aro-clusters
spec:
  clusterName: aro-hcp-cluster
  replicas: 3
  template:
    spec:
      bootstrap:
        dataSecretName: ""
      clusterName: aro-hcp-cluster
      infrastructureRef:
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
        kind: AROMachinePool
        name: aro-hcp-cluster-workers
----

. Create an `AROMachinePool` resource and define your node pool resources in `spec.resources` field. Your resource might resemble the following example:

+
[source,bash]
----
apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
kind: AROMachinePool
metadata:
  name: aro-hcp-cluster-workers
  namespace: aro-clusters
spec:
  resources:
----
```

You are not using callouts, which is great and I think it would be best to let move the info in the step. Let me know your thoughts

Author: I like this, I'll use your suggestion!

. Run the following command to apply the file:

+
[source,bash]
----
oc apply -f aro-hcp-cluster.yaml
----

+
*Note:* It might take 15 to 20 minutes for the managed cluster to provision. You can check the status of the Cluster API resources by running the following command:

+
[source,bash]
----
clusterctl describe cluster aro-hcp-cluster -n aro-clusters --show-conditions=all
----

+
You can check the status of the Azure Service Operator resources by running the following command:

+
[source,bash]
----
oc get HcpOpenShiftCluster,HcpOpenShiftClustersNodePool -n aro-clusters
----

+
You can check the Cluster API Provider for Azure controller logs by running the following command:

+
[source,bash]
----
oc logs -n multicluster-engine deployment/capz-controller-manager -f

> ----

Reviewer: I was expecting to see bullet points for the unordered list. Not sure about the alignment in my suggestion. Let me know your thoughts. This is not required. 
```suggestion
*Note:* It might take 15 to 20 minutes for the managed cluster to provision. Complete the following checks to view statuses on your cluster:

- Check the status of the Cluster API resources by running the following command:

+
[source,bash]
----
clusterctl describe cluster aro-hcp-cluster -n aro-clusters --show-conditions=all
----

- Check the status of the Azure Service Operator resources by running the following command:

+
[source,bash]
----
oc get HcpOpenShiftCluster,HcpOpenShiftClustersNodePool -n aro-clusters
----

- Check the Cluster API Provider for Azure controller logs by running the following command:

+
[source,bash]
----
oc logs -n multicluster-engine deployment/capz-controller-manager -f
----
```

Author: Each of these are steps with a command to run, I think it might better to keep them as steps instead of bullet points.

Reviewer: Ok. what do you think about adding this part to the Verification section? Starting at "You can check the status of the Azure Service Operator resources by running the following command:"

it would then just be a step in the verification to "Check the status of the Azure..."

+
[source,bash]
----
clusterctl describe cluster aro-hcp-cluster -n aro-clusters --show-conditions=all
----

+
If your cluster is ready, the output reseembles the following example:

+
[source,bash]
----
NAME              CLUSTERCLASS   AVAILABLE   CP AVAILABLE   AGE
aro-hcp-cluster                  True        True           20m
----

. Verify your infrastructure resources.

.. Check the `AROCluster` status. Run the following command:

+
[source,bash]
----
oc get arocluster aro-hcp-cluster -n aro-clusters
----

.. Check the `AROControlPlane` status. Run the following command:

+
[source,bash]
----
oc get arocontrolplane aro-hcp-cluster -n aro-clusters
----

.. Check the `AROMachinePool` status. Run the following command:

+
[source,bash]
----
oc get aromachinepool -n aro-clusters
----

. Verify your Azure Service Operator resources.

.. Check the {hcp} cluster status in Microsoft Azure. Run the following command:

+
[source,bash]
----
oc get HcpOpenShiftCluster aro-hcp-cluster -n aro-clusters
----

.. Check the node pool status. Run the following command:

+
[source,bash]
----
oc get HcpOpenShiftClustersNodePool -n aro-clusters
----

> . Get your managed cluster `kubeconfig` file. The `kubeconfig`` is automatically created when the managed cluster is ready. Run the following command:

Reviewer: ```suggestion
. Get your managed cluster `kubeconfig` file. The `kubeconfig` is automatically created when the managed cluster is ready. Run the following command:
```

  template:
    spec:
      bootstrap:
        dataSecretName: ""
      clusterName: aro-hcp-cluster
      infrastructureRef:
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
        kind: AROMachinePool
        name: aro-hcp-cluster-workers
----

. Create an `AROMachinePool` resource and define your node pool resources in `spec.resources` field. Your resource might resemble the following example:

+
[source,bash]
----
apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
kind: AROMachinePool
metadata:
  name: aro-hcp-cluster-workers
  namespace: aro-clusters
spec:
  resources:
----

. Run the following command to apply the file:

+
[source,bash]
----
oc apply -f aro-hcp-cluster.yaml
----

+
*Note:* It might take 15 to 20 minutes for the managed cluster to provision.

.Verification

. Check the status of the Cluster API resources by running the following command:

+
[source,bash]
----
clusterctl describe cluster aro-hcp-cluster -n aro-clusters --show-conditions=all
----

. Check the status of the Azure Service Operator resources by running the following command:

+
[source,bash]
----
oc get HcpOpenShiftCluster,HcpOpenShiftClustersNodePool -n aro-clusters
----

> . Cluster API Provider for Azure controller logs by running the following command:

Reviewer: ```suggestion
. Check the Cluster API Provider for Azure controller logs by running the following command:
```
